File: HW1/task3.py
import pyspark
import sys, json, operator

if __name__ == "__main__":

    # parse commandline argument
    input_file = sys.argv[1]
    output_file = sys.argv[2]
    partition_type = sys.argv[3] # only "default" or "customized"
    partition_type = partition_type == "customized"
    n_partitions = int(sys.argv[4])
    n_threshold_reviews = int(sys.argv[5])

    conf = pyspark.SparkConf().setAppName("Task2").setMaster("local[*]")
    sc = pyspark.SparkContext(conf = conf)
    sc.setLogLevel("ERROR")

    result = {}

    reviewsRDD = sc.textFile(input_file).map(lambda review: json.loads(review)).map(lambda item: (item["business_id"], 1)).cache()

    if partition_type :
        def customized_hash(x):
            return hash(x[0])
        reviewsRDD = reviewsRDD.partitionBy(n_partitions, partitionFunc=customized_hash)

    result["n_partitions"] = reviewsRDD.getNumPartitions()
    print("number of partitions: ", result["n_partitions"])

    result["n_items"] = reviewsRDD.glom().map(lambda x: len(x)).collect()
    print(result["n_items"])

    result["result"] = reviewsRDD.reduceByKey(operator.add).filter(lambda x: x[1] > n_threshold_reviews).collect()
    # print(result["result"])

    with open(output_file, 'w') as f:
        json.dump(result, f, sort_keys=True)

File: HW1/scala/task3.scala
import org.apache.spark.SparkContext
import org.apache.spark.SparkConf
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.jackson.Serialization

import java.io.{File, PrintWriter}
import scala.collection.mutable.LinkedHashMap

object task3 {
  def main(args: Array[String]) {

    // parse commandline argument
    val inputPath = args(0)
    val outputPath = args(1)
    val partitionType: Boolean = args(2).equals("customized")
    val nPartitions: Int = args(3).toInt
    val nThreshold: Int = args(4).toInt

    // set up spark
    val conf = new SparkConf().setAppName("Task3").setMaster("local[*]")
    val sc = new SparkContext(conf)
    sc.setLogLevel("ERROR")

    var result: LinkedHashMap[String, Any] = LinkedHashMap()

    def mapFunc(x: JValue): (String, Int) = {
      val str = compact(render(x \ "business_id"))
      (str.substring(1, str.length - 1), 1)
    }
    var reviewsRDD = sc.textFile(inputPath).map(line => parse(line)).map(mapFunc).cache()

    if (partitionType) {
      reviewsRDD = reviewsRDD.repartition(nPartitions)
    }

    result += ("n_partitions" -> reviewsRDD.getNumPartitions)
    println(s"number of partitions: ${result("n_partitions")}")

    result += ("n_items" -> reviewsRDD.glom().map(_.length).collect())

    result += ("result" -> reviewsRDD.reduceByKey(_+_).filter(x => x._2 > nThreshold).map(_.productIterator.toList) .collect())

    implicit val formats = org.json4s.DefaultFormats
    val writer = new PrintWriter(new File(outputPath))
    writer.write(Serialization.write(result))
    writer.close()
  }
}
